src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,121 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int INDEX_IS_NOT_PRESENT = -1;
+    private Object[] arrayList;
+    private int currentSize;
+
+    public ArrayList() {
+        arrayList = new Object[DEFAULT_CAPACITY];
+    }
+
+    private Object[] grow() {
+        return Arrays.copyOf(arrayList, currentSize + (currentSize >> 1));
+    }
+
+    private void checkSize() {
+        if (currentSize + 1 > arrayList.length) {
+            arrayList = grow();
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index >= currentSize || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    String.format("Array size is %d. Index %d out of bounds.",
+                            currentSize, index));
+        }
+    }
+
+    private int indexOfElement(T element) {
+        for (int i = 0; i < currentSize; i++) {
+            if (arrayList[i] == element
+                    || arrayList[i] != null && arrayList[i].equals(element)) {
+                return i;
+            }
+        }
+        return INDEX_IS_NOT_PRESENT;
+    }
+
+    private T removeElementByIndex(int index) {
+        T element = get(index);
+        System.arraycopy(arrayList, index + 1, arrayList, index, currentSize - index - 1);
+        currentSize--;
+        return element;
+    }
+
     @Override
     public void add(T value) {
-
+        checkSize();
+        arrayList[currentSize] = value;
+        currentSize++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > currentSize || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    String.format("Array size is %d. Index %d out of bounds.",
+                            currentSize, index));
+        }
+        checkSize();
+        System.arraycopy(arrayList, index, arrayList, index + 1, currentSize - index);
+        arrayList[index] = value;
+        currentSize++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int expectedSize = list.size() + currentSize;
+        if (expectedSize > arrayList.length) {
+            do {
+                arrayList = grow();
+            } while (expectedSize < arrayList.length);
+        }
```

Comment:
The same. The code will be more clear if the capacity growth block is highlighted separately.